scylla-cql/src/frame/value.rs

```diff
@@ -51,6 +51,81 @@ pub enum MaybeUnset<V> {
     Set(V),
 }
 
+/// Native CQL `varint` representation.
+///
+/// Represented as two's-complement binary in big-endian order.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct CqlVarint(Vec<u8>);
+
+/// Constructors from bytes
+impl CqlVarint {
+    /// Creates a [`CqlVarint`] from an array of bytes in
+    /// two's complement big-endian binary representation.
+    pub fn from_signed_bytes_be(mut digits: Vec<u8>) -> Self {
+        Self::normalize(&mut digits);
```

Comment:
@muzarski could you remind me why we are normalizing? We discussed it offline some time ago, but I don't remember it now. Please include the rationale in a comment in the code.

Reply:
We decided to normalize because `CqlValue` enum derives `PartialEq/Eq`.

It means that our new type `CqlVarint` needs to implement/derive `PartialEq/Eq` as well. The issue with this approach: take for example `[0x00, 0x01]` and `[0x01]`. If we derived `PartialEq`, the corresponding CqlVarints wouldn't be equal. However, both of these blob arrays represent the same varint value - simply 1. If we wanted them to be equal in our semantics, we would have to implement normalization anyway.

Reply:
Ok, right.

I'm thinking about an alternative: if the only issue is that we need to have `PartialEq`/`Eq` so that it works with `CqlValue`, perhaps we could skip normalization during initialization? We could normalize the values when we compare (i.e. compute a subslice which represents the normalized number and compare them).

I suppose that comparing `CqlValue`s is quite rare, while normalizing when creating the `CqlVarint` introduces some overhead to a more common path. I think this would also be a bit more consistent with the behavior of `CqlTimeuuid` in that you can pass denormalized data (in the case of CqlTimeuuid, non-timeuuid uuid).

What do you think?

Reply:
cc: @Lorak-mmk

Reply:
> I suppose that comparing `CqlValue`s is quite rare, while normalizing when creating the `CqlVarint` introduces some overhead to a more common path.

I agree. In addition, during comparison we only need to compute the subslice instead of removing elements from `Vec` - which is much cheaper.

I have an additional question regarding one edge case. What should we do with the empty vectors? Replace them with `vec![0x00]` or leave as is? The empty vector is serialized to `null`, and I'm not sure if this should be expected behaviour.

Reply:
> I have an additional question regarding one edge case. What should we do with the empty vectors? Replace them with `vec![0x00]` or leave as is? The empty vector is serialized to `null`, and I'm not sure if this should be expected behaviour.

Are you sure that they are serialized as nulls? Nulls have a different binary representation than zero-sized values: https://github.com/apache/cassandra/blob/f6707e6191a994371704f639323a31f9ecc6bcb5/doc/native_protocol_v4.spec#L229-L233

Reply:
OK, I have reproduced that and you are right. This is not a `null` value. However, the following test case (without normalization) doesn't pass:

```rust
#[tokio::test]
async fn test_cql_varint() {
    let tests = [
        vec![],
    ];

    let table_name = "cql_varint_tests";
    let session: Session = create_new_session_builder().build().await.unwrap();
    let ks = "foo_varint";

    session
        .query(
            format!(
                "CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = \
            {{'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1}}",
                ks
            ),
            &[],
        )
        .await
        .unwrap();
    session.use_keyspace(ks, false).await.unwrap();

    session
        .query(
            format!(
                "CREATE TABLE IF NOT EXISTS {} (id int PRIMARY KEY, val varint)",
                table_name
            ),
            &[],
        )
        .await
        .unwrap();

    let prepared_insert = session
        .prepare(format!(
            "INSERT INTO {} (id, val) VALUES (0, ?)",
            table_name
        ))
        .await
        .unwrap();
    let prepared_select = session
        .prepare(format!("SELECT val FROM {} WHERE id = 0", table_name))
        .await
        .unwrap();

    for test in tests {
        let cql_varint = CqlVarint::from_signed_bytes_be_slice(&test);
        session
            .execute(&prepared_insert, (&cql_varint,))
            .await
            .unwrap();

        let read_values: Vec<CqlVarint> = session
            .execute(&prepared_select, &[])
            .await
            .unwrap()
            .rows
            .unwrap()
            .into_typed::<(CqlVarint,)>()
            .map(Result::unwrap)
            .map(|row| row.0)
            .collect::<Vec<_>>();

        assert_eq!(read_values, vec![cql_varint])
    }
}
```

The error message:
```
called `Result::unwrap()` on an `Err` value: BadCqlVal { err: BadCqlType, column: 0 }
```

cqlsh output:
```cql
cqlsh> SELECT * FROM foo_varint.cql_varint_tests ;

 id | val
----+-----
  0 |    

```

However, if I change the type of `read_values` to `Vec<Option<CqlVarint>>`, and replace assertion with `assert_eq!(read_values, vec![None])` the test passes.

Reply:
The query result from the test is deserialized as `CqlValue::Empty`: https://github.com/apache/cassandra/blob/f6707e6191a994371704f639323a31f9ecc6bcb5/doc/native_protocol_v4.spec#L861-L863. Should this be expected behaviour for `CqlVarint::from_signed_bytes_be(vec![])`?

Reply:
I think this is fine. I'd still keep `CqlVarint` as a simple wrapper over a raw representation and not do any normalization at all. The protocol specification only describes normalized values, it doesn't really say what happens to the denormalized values. Maybe ScyllaDB converts them to nulls for some reason (maybe Cassandra does the same and it's about compatibility).